File: python/deepstrike/runtime/context.py
"""Host preparation adapter for the Rust Context authority."""
from __future__ import annotations

import json
from typing import Any, Callable, Protocol, TypedDict
from .evolution import ContextExecutionInput, ContextPlan, ContextState, EvaluationContextBinding


class ContextProviderPreparationRequest(TypedDict):
  effect: dict[str, Any]
  request_fingerprint: str
  provider_route: dict[str, Any]
  prompt_measurement: dict[str, Any]


class ContextPrepared(TypedDict):
  state: ContextState
  provider_route: dict[str, Any]
  prompt_measurement: dict[str, Any]
  execution_input: ContextExecutionInput
  plan: ContextPlan
  binding: EvaluationContextBinding


class ContextPreparationAdapter(Protocol):
  def prepare(self, request: ContextProviderPreparationRequest) -> ContextPrepared: ...
  def verify(self, effect: dict[str, Any], preparation: ContextPrepared) -> bool: ...


def create_context_preparation_adapter(prepare_json: Callable[[str], str], verify_json: Callable[[str], str]) -> ContextPreparationAdapter:
  class Adapter:
    def verify(self, effect: dict[str, Any], preparation: ContextPrepared) -> bool:
      return json.loads(verify_json(json.dumps({"effect": effect, "preparation": preparation}, separators=(",", ":"))))

    def prepare(self, request: ContextProviderPreparationRequest) -> ContextPrepared:
      return json.loads(prepare_json(json.dumps(request, separators=(",", ":"))))
  return Adapter()


def create_native_context_preparation_adapter() -> ContextPreparationAdapter:
  from deepstrike._kernel import context_prepare_json, context_verify_json
  return create_context_preparation_adapter(context_prepare_json, context_verify_json)

File: node/src/runtime/context.ts
import { getKernel } from "../kernel.js"
import type { RecordedPromptMeasurement, ResolvedProviderRoute } from "../providers/request-plan.js"
import type { ContextExecutionInput, ContextPlan, ContextState, EvaluationContextBinding } from "./evolution.js"

/** Host evidence completing the kernel's immutable context candidate. */
export interface ContextProviderPreparationRequest {
  readonly effect: Record<string, unknown>
  readonly request_fingerprint: string
  readonly provider_route: ResolvedProviderRoute
  readonly prompt_measurement: RecordedPromptMeasurement
}
export interface ContextPrepared {
  readonly state: ContextState
  readonly provider_route: ResolvedProviderRoute
  readonly prompt_measurement: RecordedPromptMeasurement
  readonly execution_input: ContextExecutionInput
  readonly plan: ContextPlan
  readonly binding: EvaluationContextBinding
}
export type ContextPrepareJson = (request: string) => string
export type ContextVerifyJson = (request: string) => string

/** All validation and content addressing are delegated to the Rust authority. */
export function createContextPreparationAdapter(prepareJson: ContextPrepareJson, verifyJson: ContextVerifyJson) {
  return {
    verify(effect: Record<string, unknown>, preparation: ContextPrepared): boolean {
      return JSON.parse(verifyJson(JSON.stringify({ effect, preparation }))) as boolean
    },
    prepare(request: ContextProviderPreparationRequest): ContextPrepared {
      return JSON.parse(prepareJson(JSON.stringify(request))) as ContextPrepared
    },
  }
}

export function createNativeContextPreparationAdapter() {
  const kernel = getKernel()
  return createContextPreparationAdapter(kernel.contextPrepareJson, kernel.contextVerifyJson)
}
